pkg/archive/archive.go

```diff
@@ -124,15 +125,66 @@ func IsArchivePath(path string) bool {
 	return err == nil
 }
 
+const (
+	ZstdMagicSkippableStart = 0x184D2A50
+	ZstdMagicSkippableMask  = 0xFFFFFFF0
+)
+
+var (
+	Bzip2Magic = []byte{0x42, 0x5A, 0x68}
+	GzipMagic  = []byte{0x1F, 0x8B, 0x08}
+	XzMagic    = []byte{0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00}
+	ZstdMagic  = []byte{0x28, 0xb5, 0x2f, 0xfd}
+)
+
+type matcher = func([]byte) bool
+
+func magicNumberMatcher(m []byte) matcher {
+	return func(source []byte) bool {
+		return bytes.HasPrefix(source, m)
+	}
+}
+
+// ZstDetector detects zstd compression algorithm.
+// Zstandard compressed data is made of one or more frames.
+// There are two frame formats defined by Zstandard: Zstandard frames and Skippable frames.
+// See https://tools.ietf.org/id/draft-kucherawy-dispatch-zstd-00.html#rfc.section.2 for more details.
+func ZstDetector(buf []byte) bool {
```

Comment:
Wondering if this is something to contribute to github.com/klauspost/compress/zstd (so that other consumers of that package can also implement the same detection)

Reply:
I think that library is probably aimed at users who know what compression algorithm they should choose, and it is more concerned with the implementation of the specified algorithm, so detecting the compression algorithm is not its main function.

Reply:
Yeah, my train of thought here is that;

- I know at least multiple runtimes / container-related projects that will likely (?) have to implement this detection (buildkit, containerd, moby, cri-o?)
- Afaik, most (if not all) of those currently use the klauspost module
- The klauspost module also has definitions for these magic numbers, and (from a quick look) looks to have an implementation to detect;

https://github.com/klauspost/compress/blob/f118b5f6f7e720b1f1c1464bb11904261c06f2f3/zstd/framedec.go#L57-L60
https://github.com/klauspost/compress/blob/baa1f1e42d738c71ea4a18b673ebacee0ce520b8/zstd/decodeheader.go#L79


So _perhaps_ it could be within scope of that project to have each compression export a "detection" function (and possibly a utility package that allows to detect compression for all, but that could be left to consumers)

Of course, an alternative could be to have a small module somewhere that can be used by other the projects mentioned, but having it in the klauspost project would help making sure the detection doesn't diverge.


@klauspost @giuseppe @tonistiigi any thoughts? (it's just an idea; not a show-stopper)

Reply:
fwiw there is already a [frame header decoder](https://pkg.go.dev/github.com/klauspost/compress/zstd#Header.Decode) function where you can check if the [HeaderMaxSize = 17](https://pkg.go.dev/github.com/klauspost/compress/zstd#pkg-constants) bytes is a valid zstd (skippable) header.